### PR TITLE
Breaking: validate parser options (fixes #384)

### LINKED ISSUE #384: Option validation
It could be useful to add option validation to espree and make sure that incoming options match a schema.

Are there any downsides? Do we want to allow variadic options? Should we make this validation itself opt-in?

(Inspired by eslint/eslint#10479.)

### PR DESCRIPTION
**Pull request by aladdin-add**
it will throw an error if any of the conditions is true:

+ ecmaVersion is invalid
+ sourceType is invalid
+ ecmaVersion < 6 & sourceType: "module"

is there some more conditions need to be covered?

### CODE REVIEW

File: lib/espree.js [hunk 1]
```diff
@@ -36,11 +38,39 @@ function normalizeEcmaVersion(ecmaVersion) {
...
+ * @param {Object} options the parser options to normalize
+ * @throws {Error} throw an error if found invalid option.
+ * @returns {Object} normalized options
+ */
+function normalizeOptions(options) {
+    const ecmaVersion = normalizeEcmaVersion(options.ecmaVersion);
+    const sourceType = normalizeSourceType(options.sourceType);
+
+    if (sourceType === "module" && ecmaVersion < 6) {
+        throw new Error("sourceType 'module' is not supported when ecmaVersion < 6!");
```
Comment (platinumazure) [inline lib/espree.js]: I don't think we should use an exclamation point here-- it's unnecessary, and could potentially confuse people.

```suggestion
        throw new Error("sourceType 'module' is not supported when ecmaVersion < 6");
```

File: tests/lib/ecma-version.js [hunk 2]
```diff
@@ -141,6 +141,30 @@ describe("ecmaVersion", () => {
             }, /Invalid ecmaVersion/);
         });
 
+        it("Should throw error using invalid year", () => {
```
Comment (platinumazure) [inline tests/lib/ecma-version.js]: This test case has the same name as the previous test case. Is that intentional?

One possible suggestion (feel free to disregard):

```suggestion
        it("Should throw error when non-numeric year is provided", () => {
```

File: lib/espree.js [hunk 3]
```diff
@@ -36,11 +38,41 @@ function normalizeEcmaVersion(ecmaVersion) {
                 return version;
 
             default:
```
Comment (not-an-aardvark) [inline lib/espree.js]:
```suggestion
            // no default
```

File: lib/espree.js [hunk 4]
```diff
@@ -36,11 +38,41 @@ function normalizeEcmaVersion(ecmaVersion) {
...
+ * @returns {Object} normalized options
+ */
+function normalizeOptions(options) {
+    const ecmaVersion = normalizeEcmaVersion(options.ecmaVersion);
+    const sourceType = normalizeSourceType(options.sourceType);
+    const ranges = options.range === true;
+    const locations = options.loc === true;
+
+    if (sourceType === "module" && ecmaVersion < 6) {
+        throw new Error("sourceType 'module' is not supported when ecmaVersion < 6.");
```
Comment (not-an-aardvark) [inline lib/espree.js]: It seems like this error will often be shown to some ESLint users. Should we make it more instructive about how to fix the problem?

```suggestion
        throw new Error("sourceType 'module' is not supported when ecmaVersion < 2015. Consider adding `{ ecmaVersion: 2015 }` to the parser options.");
```

**Comment by not-an-aardvark** [comment]
Does this also fix https://github.com/eslint/eslint/issues/9687, or are more changes in ESLint needed to fix that?

**Comment by aladdin-add** [comment]
> Does this also fix eslint/eslint#9687, or are more changes in ESLint needed to fix that?

we need to upgrade espree, since this will be in a major release. not sure some more, like adding a test?

**Comment by simonbrunel** [comment]
So does that mean ESLint shouldn't be used in projects targeting `ecmaVersion < 2015`, while not using any transpiler but for which sources rely on ES6 import/export bundler (e.g. rollup). I don't think ESLint should enforce the `ecmaVersion >= 2015`  when `sourceType: module` since this should be an acceptable use case:

```javascript
module.exports = {
    parserOptions: {
        ecmaVersion: 5,       // disable ES6 features (const, let, etc.)
        sourceType: module    // allow ESM import/export syntax
    }
}
```

**Comment by not-an-aardvark** [comment]
@simonbrunel To be clear, ESLint already does not support that case. The config you showed currently has the effect of setting `ecmaVersion: 2015` despite the `ecmaVersion: 5` value provided in the config. This change just makes the error more obvious to avoid setting any misleading expectations.

**Comment by simonbrunel** [comment]
@not-an-aardvark yes I know, but these kind of projects exist and I don't think ESLint should exclude them.

**Comment by simonbrunel** [comment]
What I mean is instead of setting `ecmaVersion` to `2015` automatically or trigger an error if it doesn't match when `sourceType: module`, it may be better to make a distinction between the scope of these 2 options and allow projects that use ESM import/export to disable other ES6 features.

**Comment by aladdin-add** [comment]
you can use [eslint-plugin-es](https://www.npmjs.com/package/eslint-plugin-es).

**Comment by simonbrunel** [comment]
@aladdin-add thanks, I will give it a try!

**Pull request merged**